[cssom] Don't serialize shorthand if longhands have different importance

Loirooriol opened the pull request:
#### dc7aed5d4dc6e98d4f0832ceb97c02a085c93379
<pre>
[cssom] Don&apos;t serialize shorthand if longhands have different importance
<a href="https://bugs.webkit.org/show_bug.cgi?id=247771">https://bugs.webkit.org/show_bug.cgi?id=247771</a>

Reviewed by Darin Adler.

When some longhand is !important and another isn&apos;t, then the shorthand
should serialize as empty string. But most shorthands weren&apos;t checking.

This change brings consistency and matches Gecko, Blink and the spec:
<a href="https://drafts.csswg.org/cssom/#dom-cssstyledeclaration-getpropertyvalue">https://drafts.csswg.org/cssom/#dom-cssstyledeclaration-getpropertyvalue</a>

Test: imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks.html

The test still has several failures, but the check for longhands with
different importance is now passing.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt: Added.
Test expectation.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks.html: Added.
New test.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::commonShorthandChecks const):
This method replaces shorthandHasVariableReference. It has a different
name since it now it also checks for importance, and in the future it
should check for even more things so that the test fully passes.

(WebCore::StyleProperties::getPropertyValue const):
Use commonShorthandChecks instead of the old shorthandHasVariableReference.
Assert that all shorthands have serialization logic.

(WebCore::StyleProperties::get2Values const):
Important flags are now checked in commonShorthandChecks().

(WebCore::StyleProperties::get4Values const):
Important flags are now checked in commonShorthandChecks().

(WebCore::StyleProperties::getCommonValue const):
Important flags are now checked in commonShorthandChecks().

(WebCore::StyleProperties::borderImagePropertyValue const):
Important flags are now checked in commonShorthandChecks().

(WebCore::StyleProperties::shorthandHasVariableReference const): Deleted.
Renamed to commonShorthandChecks.

* Source/WebCore/css/StyleProperties.h:
Rename shorthandHasVariableReference to commonShorthandChecks.

Canonical link: <a href="https://commits.webkit.org/256774@main">https://commits.webkit.org/256774@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9f7451aedbafca84f7ebe37db0af17994afebf7e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96711 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5976 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29803 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106237 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166545 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100695 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6187 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34711 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89093 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102950 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102387 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/4623 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83322 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31587 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86463 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88336 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74516 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19819 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21240 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4703 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4796 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43765 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1237 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40520 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -177,35 +177,44 @@ MutableStyleProperties::MutableStyleProperties(const StyleProperties& other)
+        // Don't serialize the shorthand if longhands have different importance.
+        auto property = propertyAt(propertyIndex);
+        if (!importance)
+            importance.emplace(property.isImportant());
```
There is no reason to use `emplace` instead of `=`,here. For the type `bool` there is no move optimization needed.

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -177,35 +177,44 @@ MutableStyleProperties::MutableStyleProperties(const StyleProperties& other)
+        else if (*importance != property.isImportant())
+            return String();
+
+        auto cssPropertyValue = property.value();
```
No need for such a long name for this local. Just call it `value` and given the scope (a loop through properties) it is clear.

Loirooriol replied:
Done

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -177,35 +177,44 @@ MutableStyleProperties::MutableStyleProperties(const StyleProperties& other)
-            shorthandValue = downcast<CSSPendingSubstitutionValue>(* getPropertyCSSValue(shorthand.properties()[0])).shorthandValue().cssText();
-            return true;
-        }
+    if (!shorthand.length())
```
Not new to your changes but no need for a special case, the code below does the right thing for length zero.

Loirooriol replied:
Yes, but serializing longhands as empty string allows separating the `ASSERT_NOT_REACHED()` in the default case of the switch. I think it's good to separate the logic for longhands and shorthands.

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -226,11 +235,8 @@ String StyleProperties::getPropertyValue(CSSPropertyID propertyID) const
-        if (shorthandHasVariableReference(propertyID, shorthandValue))
-            return shorthandValue;
-    }
+    if (std::optional<String> shorthandValue = commonShorthandChecks(propertyID))
```
Since `String` has a distinct null and empty value I suggest not bothering with `optional`. The function should return `emptyString()` for cases that should serialize as the empty string and null string if there is no special consideration. The code can be:
```
if (auto result = commonShorthandChecks(propertyID); !result.isNull())
    return result;
```

Loirooriol replied:
But this seems more risky, e.g. in `StyleProperties::asTextInternal()` it breaks

```cpp
            value = getPropertyValue(shorthandPropertyID);
            if (!value.isNull()) {
```

There are lots of callers, most of which return the value to even more callers, so it seems easy to miss something. This seems safer:

```cpp
if (auto result = commonShorthandChecks(propertyID); !result.isNull())
    return result.isEmpty() ? nullString() : result;
```

darinadler replied:
Fine for your change now, but not a great situation.

Now that you pointed this out I think the code is fragile right now. Functions like `fontValue` are returning empty strings to mean nit serializable, not null string. We should probably change `asTextInternal` to check for empty, not null, or find all the places using empty string and change them to consistently use null. Maybe even find a place to assert that.

The primary use of `getPropertyValue`, where it’s exposed to the bindings, will treat empty string and null string the same. If we want to use null string to mean something special internally, we should figure out what that is. Maybe null should mean “property literally does not exist” in some stronger sense than “serializes as empty string”.

Loirooriol replied:
Luckily `asTextInternal()` doesn't attempt to use `CSSPropertyFont`, but it seems fragile indeed.

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -177,35 +177,44 @@ MutableStyleProperties::MutableStyleProperties(const StyleProperties& other)
+
+        // Don't serialize the shorthand if longhands have different importance.
+        auto property = propertyAt(propertyIndex);
+        if (!importance)
```
In my version I wrote this:
```
bool isImportant = property.isImportant()
if (importance.value_or(isImportant) != isImportant)
    return emptyString();
Importance = isImportant;
```
You may prefer the way you wrote it, but I thought I’d say anyway.

Loirooriol replied:
Switched to that.

darinadler reviewed (approved):
This is great. I have a very similar change in my local tree that I haven’t had time to complete yet, and it’s good to get this in!

My comments are all optional code style thoughts.

Loirooriol committed:
[cssom] Don't serialize shorthand if longhands have different importance
https://bugs.webkit.org/show_bug.cgi?id=247771

Reviewed by Darin Adler.

When some longhand is !important and another isn't, then the shorthand
should serialize as empty string. But most shorthands weren't checking.

This change brings consistency and matches Gecko, Blink and the spec:
https://drafts.csswg.org/cssom/#dom-cssstyledeclaration-getpropertyvalue

Test: imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks.html

The test still has several failures, but the check for longhands with
different importance is now passing.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt: Added.
Test expectation.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks.html: Added.
New test.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::commonShorthandChecks const):
This method replaces shorthandHasVariableReference. It has a different
name since it now it also checks for importance, and in the future it
should check for even more things so that the test fully passes.

(WebCore::StyleProperties::getPropertyValue const):
Use commonShorthandChecks instead of the old shorthandHasVariableReference.
Assert that all shorthands have serialization logic.

(WebCore::StyleProperties::get2Values const):
Important flags are now checked in commonShorthandChecks().

(WebCore::StyleProperties::get4Values const):
Important flags are now checked in commonShorthandChecks().

(WebCore::StyleProperties::getCommonValue const):
Important flags are now checked in commonShorthandChecks().

(WebCore::StyleProperties::borderImagePropertyValue const):
Important flags are now checked in commonShorthandChecks().

(WebCore::StyleProperties::shorthandHasVariableReference const): Deleted.
Renamed to commonShorthandChecks.

* Source/WebCore/css/StyleProperties.h:
Rename shorthandHasVariableReference to commonShorthandChecks.

Canonical link: https://commits.webkit.org/256774@main
